pep-9999.txt

```diff
@@ -0,0 +1,171 @@
+PEP: 9999
+Title: Python local packages directory [We need to find a proper name]
+Version: $Revision$
+Last-Modified: $Date$
+Author: Kushal Das <mail@kushaldas.in>, Steve Dower <steve.dower@python.org>,
+        Donald Stufft <donald@stufft.io>, Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 16-May-2018
+Python-Version: 3.8
+
+
+
+Abstract
+========
+
+This PEP proposes to add to Python a mechanism to have a ``__pypackages__``
+directory  per project to have its own independent set of installed packages.
+This will share the standard library with the base installed Python. This will
+avoid the steps to create or activate or deactivate "virtual environments". Any
+Python scripts will use the ``__pypackages__`` from the base directory of the 
+script.
+
+
+
+Motivation
+==========
+
+Python virtual environments have become an essential part of development and
+teaching workflow in the community, but at the same time, they create a
+barrier to entry for many folks. The following are few examples of the issues
+people get into while being introduced to Python (or programming for the first
+time).
+
+- How virtual environments work is a lot of information for anyone new. It takes
+  a lot of extra time and effort to explain them.
+
+- Different platforms and shell environments require different sets of commands
+  to work on the virtual environments. Any workshop or teaching environment with
+  people coming with different operating systems installed on the laptops create a
+  lot of confusion among the participants.
+
+- Virtual environments need to be activated on each opened terminal, if someone
+  creates/opens a new terminal, that by default does not get the same environment
+  as in a previous terminal with virtual environment activated.
+
+
+Specification
+=============
+
+When the Python binary is executed, it attempts to determine its
+prefix (which it stores in ``sys.prefix``), which is then used to find
+the standard library and other key files, and by the ``site`` module
+to determine the location of the site-package directories.  Currently
+the prefix is found (assuming ``PYTHONHOME`` is not set) by first
+walking up the filesystem tree looking for a marker file (``os.py``)
+that signifies the presence of the standard library, and if none is
+found, falling back to the build-time prefix hard coded in the binary.
+
+This PEP proposes to add a new step in this process. If a ``__pypackages__``
+directory is found in the current working directory, then it will be included in
```

Comment:
The npm/node.js solution that is similar to this doesn't just check the local directory, but it walks up the directory tree looking for it as well. I think that is an important feature to have, otherwise you'll end up with confusion if people ``cd`` into a sub directory.

Reply:
That seems like a pretty esoteric use case to me? It'd require someone to have write permissions to a higher level directory but not also be able to just take permissions from the lower level directory. Is there a threat model for when someone is able to drop ``../../__pypackages__/``, but can't influence the ``site-packages`` another way?

Reply:
In what circumstances is the current working directory found?  From the explanation kushaldas gave of the examples, it seems like most of them should actually be found next to the script being invoked.  IIRC, python -m MODULE is the only case where __pypackages__ would be searched for in the current working directory?

(If so, this owrding should be amended).

Reply:
For what it's worth (and this is sort of related to the isolation question below), but not only does node.js walk up the directory tree, it also doesn't stop at the first ``node_modules/`` it finds. It would effectively look like:

```python-console
>>> import pprint
>>> import sys
>>> pprint.pprint(sys.path)
['',
 '/Users/dstufft/.pyenv/versions/3.6.2/lib/python36.zip',
 '/Users/dstufft/.pyenv/versions/3.6.2/lib/python3.6',
 '/Users/dstufft/.pyenv/versions/3.6.2/lib/python3.6/lib-dynload',

 # The new paths added as a result of this, assuming our current context is
 # ~/projects/example/sub-example. Of course site.py wouldn't add any of these
 # directories that don't exist, much like it does for user site-packages and such,
 # so in practice this would typically end up with 0 or 1 entries from this set,
 # but it enables more advanced use cases such as sharing a set of dependencies at
 # the top level of a package, and override them at a sub package level.
 '/Users/dstufft/projects/example/sub-example/__pypackages__/Lib/site-packages',
 '/Users/dstufft/projects/example/__pypackages__/Lib/site-packages',
 '/Users/dstufft/projects/__pypackages__/Lib/site-packages',
 '/Users/dstufft/__pypackages__/Lib/site-packages',
 '/Users/__pypackages__/Lib/site-packages',
 '/__pypackages__/Lib/site-packages',

 '/Users/dstufft/.local/lib/python3.6/site-packages',
 '/Users/dstufft/.pyenv/versions/3.6.2/lib/python3.6/site-packages']
```

I know that I would find the behavior without parent recursing super inconvenient, particularly coming from the venv world where I'm already used to be able to cd all around my project directory without suddenly losing access to my installed dependencies. I believe this is also the way that pipenv works (walking up the directory tree to locate the virtual environment), so I think that the prior art in general points towards walking.

Reply:
> Is there a threat model for when someone is able to drop ../../__pypackages__/, but can't influence the site-packages another way?

Just off the top of my head: on a Unix-y system, running scripts from `/tmp/` or a subdirectory of it would open you to attacks from whoever owns `/tmp/__pypackages__`, and that can be anyone.

I really, *really* like the overlaying/walking behavior from Node, and I think it should be possible to meaningfully add it to Python, but the security side will need some careful thought.

Reply:
I wonder if it makes sense to load a ``__pypackages__ directory`` that is owned by a different user?

This would be a different behavior from the typical ``site-packages`` directory loading, but since this is for a sort of lightweight environment, it may make sense to limit it to only looking in a ``__pypackages__`` directory that is owned by the current user, possibly even terminating the upwards walk as soon as we find a directory that isn't owned by the current user. 

This would limit to some degree the usefulness of the feature, but I think the use cases it excludes are edge cases or are also already reasonably handled by our global or user site-packages directories. That would then mean in order to attack a user, the attacker would need to be able to write files *as that user*, which if they can do that, they can already attack the user in many other ways that it's not worth it to worry about.